### ISSUE TITLE
Dead lock in docker events

### ISSUE THREAD

**Issue by gambol99** [fsouza/go-dockerclient#203]
I keep hitting a dead lock when listening to docker events; a section of the code below

	   docker_events := make(chan *docker.APIEvents)
	    	go func() {
	    		/* step: add our channel as an event listener for docker events */
	    		if err := r.Docker.AddEventListener(docker_events); err != nil {
	    			glog.Fatalf("Unable to register docker events listener, error: %s", err)
	    			return
	    		}
	    		/* step: start the event loop and wait for docker events */
	    		glog.V(5).Infof("Entering into the docker events loop")
	    		for {
	    			select {
	    			case event := <-docker_events:
	    				glog.V(4).Infof("Received docker event status: %s, id: %s", event.Status, event.ID)
	    				switch event.Status {
	    				case DOCKER_START:
	    					go r.ProcessDockerCreation(event.ID, channel)
	    				case DOCKER_DESTROY:
	    					go r.ProcessDockerDestroy(event.ID, channel)
	    				}
	    			}
	    		}
	    		glog.Errorf("Exitting the docker events loop")
	    	}()

Eventually the loop will stop listening to docker events - performing a kill -6 PID

    goroutine 22 [semacquire]:
    sync.runtime_Semacquire(0xc20805215c)
      /usr/lib/golang/src/pkg/runtime/sema.goc:199 +0x30
    sync.(*RWMutex).RLock(0xc208052150)
      /usr/lib/golang/src/pkg/sync/rwmutex.go:36 +0x5e
    github.com/fsouza/go-dockerclient.(*eventMonitoringState).isEnabled(0xc208052150, 0xc20815c400)
      /home/jest/go/src/github.com/gambol99/embassy/Godeps/_workspace/src/github.com/fsouza/go-dockerclient/event.go:219 +0x33
    github.com/fsouza/go-dockerclient.(*eventMonitoringState).monitorEvents(0xc208052150, 0xc2080a6000)
      /home/jest/go/src/github.com/gambol99/embassy/Godeps/_workspace/src/github.com/fsouza/go-dockerclient/event.go:173 +0xf7
    created by github.com/fsouza/go-dockerclient.(*eventMonitoringState).enableEventMonitoring
      /home/jest/go/src/github.com/gambol99/embassy/Godeps/_workspace/src/github.com/fsouza/go-dockerclient/event.go:148 +0x113
    
    goroutine 36 [semacquire]:
    sync.runtime_Semacquire(0xc20805215c)
      /usr/lib/golang/src/pkg/runtime/sema.goc:199 +0x30
    sync.(*RWMutex).RLock(0xc208052150)
      /usr/lib/golang/src/pkg/sync/rwmutex.go:36 +0x5e
    github.com/fsouza/go-dockerclient.(*eventMonitoringState).isEnabled(0xc208052150, 0x6ae800)
      /home/jest/go/src/github.com/gambol99/embassy/Godeps/_workspace/src/github.com/fsouza/go-dockerclient/event.go:219 +0x33
    github.com/fsouza/go-dockerclient.func·004(0xc2080442d0, 0xc20803c2c0)
      /home/jest/go/src/github.com/gambol99/embassy/Godeps/_workspace/src/github.com/fsouza/go-dockerclient/event.go:302 +0x308
    created by github.com/fsouza/go-dockerclient.(*Client).eventHijack
      /home/jest/go/src/github.com/gambol99/embassy/Godeps/_workspace/src/github.com/fsouza/go-dockerclient/event.go:307 +0x4a6
    
    goroutine 99 [semacquire]:
    sync.runtime_Semacquire(0xc208052158)
      /usr/lib/golang/src/pkg/runtime/sema.goc:199 +0x30
    sync.(*RWMutex).Lock(0xc208052150)
      /usr/lib/golang/src/pkg/sync/rwmutex.go:83 +0xa3
    github.com/fsouza/go-dockerclient.(*eventMonitoringState).updateLastSeen(0xc208052150, 0xc2080b2d80)
      /home/jest/go/src/github.com/gambol99/embassy/Godeps/_workspace/src/github.com/fsouza/go-dockerclient/event.go:242 +0x32
    created by github.com/fsouza/go-dockerclient.(*eventMonitoringState).monitorEvents
      /home/jest/go/src/github.com/gambol99/embassy/Godeps/_workspace/src/github.com/fsouza/go-dockerclient/event.go:186 +0x22a
    
    goroutine 98 [semacquire]:
    sync.runtime_Semacquire(0xc20805215c)
      /usr/lib/golang/src/pkg/runtime/sema.goc:199 +0x30
    sync.(*RWMutex).RLock(0xc208052150)
      /usr/lib/golang/src/pkg/sync/rwmutex.go:36 +0x5e
    github.com/fsouza/go-dockerclient.(*eventMonitoringState).noListeners(0xc208052150, 0x0)
      /home/jest/go/src/github.com/gambol99/embassy/Godeps/_workspace/src/github.com/fsouza/go-dockerclient/event.go:213 +0x33
    github.com/fsouza/go-dockerclient.(*eventMonitoringState).sendEvent(0xc208052150, 0xc2080b2d80)
      /home/jest/go/src/github.com/gambol99/embassy/Godeps/_workspace/src/github.com/fsouza/go-dockerclient/event.go:230 +0xe1
    created by github.com/fsouza/go-dockerclient.(*eventMonitoringState).monitorEvents
      /home/jest/go/src/github.com/gambol99/embassy/Godeps/_workspace/src/github.com/fsouza/go-dockerclient/event.go:185 +0x205

On every occurrence of the event, the stack above it always the same ... X in RLock() and the updateLastSeen() trying to Lock(). Looking at the code (added some debug lines)

	event.go
    line 177: case ev, ok := <-eventState.C:
    			fmt.Fprintf(os.Stdout,"docker() event: %V\n", ev )
    			if !ok {
    				return
    			}
    			if ev == EOFEvent {
    				eventState.closeListeners()
    				eventState.terminate()
    				return
    			}
    			go eventState.sendEvent(ev)
    			go eventState.updateLastSeen(ev)

....

    func (eventState *eventMonitoringState) sendEvent(event *APIEvents) {
    	fmt.Fprintf(os.Stdout, "sendEvent() about to lock\n")
    	eventState.RLock()
    	fmt.Fprintf(os.Stdout,"sendEvent() locked\n")
    	defer eventState.RUnlock()
    	fmt.Fprintf(os.Stdout, "sendEvent() adding event state\n")
    	eventState.Add(1)
    	defer eventState.Done()
    	fmt.Fprintf(os.Stdout, "sendEvent() added event state, eventState: %V", eventState )
    	if eventState.isEnabled() { <-- the issue perhaps?
          Fprintf(os.Stdout, "sendEvent() is enabled")

    func (eventState *eventMonitoringState) isEnabled() bool {
		eventState.RLock()
		defer eventState.RUnlock()
		return eventState.enabled
    }

    func (eventState *eventMonitoringState) updateLastSeen(e *APIEvents) {
    	fmt.Fprintf(os.Stdout,"updateLastSeen() about to lock\n")
    	eventState.Lock()
    	fmt.Fprintf(os.Stdout,"updateLastSeen() locked\n")
    	defer eventState.Unlock()
    	if atomic.LoadInt64(eventState.lastSeen) < e.Time {
    		atomic.StoreInt64(eventState.lastSeen, e.Time)
    	}
    	fmt.Fprintf(os.Stdout,"updateLastSeen() exitting\n")
    }

Output from a deadlock logging

    	docker() event: &{%!V(string=create) %!V(string=aa257197e9d1416e9c8883edf69c83de89e4b30f6e7deaab3a949d09f918e6bb) %!V(string=centos:centos7) %!V(int64=1421163166)}
    sendEvent() about to lock
    updateLastSeen() about to lock
    sendEvent() locked
    sendEvent() adding event state
    sendEvent() added event state
    .... nothing afterwards

If i'm not mistaken (https://golang.org/src/sync/rwmutex.go) .. The Lock() in updateLastSeen() will block the isEnabled() from running, as it also requires a read lock. Thus blocking sendEvent() from finishing, performing an RUnlock() on the defer and allowing updateLastSeen()  to proceed; halting any new events?? ... or am I incorrect?

	70	// Lock locks rw for writing.
    71	// If the lock is already locked for reading or writing,
    72	// Lock blocks until the lock is available.
    73	// To ensure that the lock eventually becomes available,
    74	// a blocked Lock call excludes new readers from acquiring
    75	// the lock.
    76	func (rw *RWMutex) Lock() {

**Comment by blalor** [comment]
Pretty sure this is related to (or, rather, the cause of) #202.  I didn't know about the `kill -ABRT <pid>` trick; that's extremely helpful.  My stack trace is similar:

```
goroutine 36 [semacquire, 5128 minutes]:
sync.runtime_Semacquire(0xc20803e62c)
	/usr/lib/golang/src/pkg/runtime/sema.goc:199 +0x30
sync.(*RWMutex).RLock(0xc20803e620)
	/usr/lib/golang/src/pkg/sync/rwmutex.go:36 +0x5e
github.com/fsouza/go-dockerclient.(*eventMonitoringState).isEnabled(0xc20803e620, 0x696b00)
	github.com/fsouza/go-dockerclient/event.go:219 +0x33
github.com/fsouza/go-dockerclient.func·004(0xc208063320, 0xc2080b64d0)
	github.com/fsouza/go-dockerclient/event.go:302 +0x308
created by github.com/fsouza/go-dockerclient.(*Client).eventHijack
	github.com/fsouza/go-dockerclient/event.go:307 +0x4a6

goroutine 21 [semacquire, 5128 minutes]:
sync.runtime_Semacquire(0xc20803e62c)
	/usr/lib/golang/src/pkg/runtime/sema.goc:199 +0x30
sync.(*RWMutex).RLock(0xc20803e620)
	/usr/lib/golang/src/pkg/sync/rwmutex.go:36 +0x5e
github.com/fsouza/go-dockerclient.(*eventMonitoringState).isEnabled(0xc20803e620, 0xc20810b000)
	github.com/fsouza/go-dockerclient/event.go:219 +0x33
github.com/fsouza/go-dockerclient.(*eventMonitoringState).monitorEvents(0xc20803e620, 0xc208064600)
	github.com/fsouza/go-dockerclient/event.go:173 +0xf7
created by github.com/fsouza/go-dockerclient.(*eventMonitoringState).enableEventMonitoring
	github.com/fsouza/go-dockerclient/event.go:148 +0x113

goroutine 47 [semacquire, 5128 minutes]:
sync.runtime_Semacquire(0xc20803e62c)
	/usr/lib/golang/src/pkg/runtime/sema.goc:199 +0x30
sync.(*RWMutex).RLock(0xc20803e620)
	/usr/lib/golang/src/pkg/sync/rwmutex.go:36 +0x5e
github.com/fsouza/go-dockerclient.(*eventMonitoringState).noListeners(0xc20803e620, 0x0)
	github.com/fsouza/go-dockerclient/event.go:213 +0x33
github.com/fsouza/go-dockerclient.(*eventMonitoringState).sendEvent(0xc20803e620, 0xc208086200)
	github.com/fsouza/go-dockerclient/event.go:230 +0xe1
created by github.com/fsouza/go-dockerclient.(*eventMonitoringState).monitorEvents
	github.com/fsouza/go-dockerclient/event.go:185 +0x205

goroutine 48 [semacquire, 5128 minutes]:
sync.runtime_Semacquire(0xc20803e628)
	/usr/lib/golang/src/pkg/runtime/sema.goc:199 +0x30
sync.(*RWMutex).Lock(0xc20803e620)
	/usr/lib/golang/src/pkg/sync/rwmutex.go:83 +0xa3
github.com/fsouza/go-dockerclient.(*eventMonitoringState).updateLastSeen(0xc20803e620, 0xc208086200)
	github.com/fsouza/go-dockerclient/event.go:242 +0x32
created by github.com/fsouza/go-dockerclient.(*eventMonitoringState).monitorEvents
	github.com/fsouza/go-dockerclient/event.go:186 +0x22a
```

I'm building with fsouza/go-dockerclient@bc54f3734b9b098e24680d5ad686f602bc152df4.

**Comment by blalor** [comment]
There's an awful lot of locking going on around here, much of it superfluous, I think. Particularly with the atomic calls around `lastSeen`.  I can't see any reason for `updateLastSeen` to be its own goroutine, and I think that's part of the cause of the deadlock.  I realize much of this is to deal with seamless reconnects, but it would be much simpler to just notify the client that the connection to the server died.

**Comment by fsouza** [comment]
Closing this, thanks to @blalor fix on #206.
